core/src/forwarding_stage.rs

```diff
@@ -45,95 +42,178 @@ mod packet_container;
-        let pkts = batch.iter().map(|(bytes, addr)| (bytes, addr));
-        let _res = batch_send(&self.udp_socket, pkts);
-        batch.clear();
+impl ForwardingClient for VoteClient {
+    fn update_address(&mut self) -> bool {
+        self.current_address = self.forward_address_getter.get_vote_forwarding_addresses();
+        self.current_address.is_some()
     }
-}
 
-/// Forwards packets to current/next leader.
-pub struct ForwardingStage<F: ForwardAddressGetter> {
-    receiver: Receiver<(BankingPacketBatch, bool)>,
-    packet_container: PacketContainer,
+    fn send_transactions_in_batch(&self, wire_transactions: Vec<Vec<u8>>) {
+        assert!(
+            self.current_address.is_some(),
+            "current_address should be updated before send_batch call."
+        );
+        let batch_with_addresses = wire_transactions
+            .iter()
+            .map(|bytes| (bytes, self.current_address.unwrap()));
+        let _res = batch_send(&self.udp_socket, batch_with_addresses);
+    }
+}
 
-    root_bank_cache: RootBankCache,
-    forward_address_getter: F,
+#[derive(Clone)]
+struct ConnectionCacheClient {
     connection_cache: Arc<ConnectionCache>,
-    vote_client: VoteClient,
-    data_budget: DataBudget,
-
-    metrics: ForwardingStageMetrics,
+    forward_address_getter: ForwardAddressGetter,
+    current_address: Option<SocketAddr>,
 }
 
-impl<F: ForwardAddressGetter> ForwardingStage<F> {
-    pub fn spawn(
-        receiver: Receiver<(BankingPacketBatch, bool)>,
+impl ConnectionCacheClient {
+    fn new(
         connection_cache: Arc<ConnectionCache>,
-        root_bank_cache: RootBankCache,
-        forward_address_getter: F,
-        data_budget: DataBudget,
-    ) -> JoinHandle<()> {
-        let forwarding_stage = Self::new(
-            receiver,
+        forward_address_getter: ForwardAddressGetter,
+    ) -> Self {
+        Self {
             connection_cache,
-            root_bank_cache,
             forward_address_getter,
-            data_budget,
+            current_address: None,
+        }
+    }
+}
+
+impl ForwardingClient for ConnectionCacheClient {
+    fn update_address(&mut self) -> bool {
+        self.current_address = self
+            .forward_address_getter
+            .get_non_vote_forwarding_addresses(1, self.connection_cache.protocol())[0];
+        self.current_address.is_some()
+    }
+
+    fn send_transactions_in_batch(&self, wire_transactions: Vec<Vec<u8>>) {
+        assert!(
+            self.current_address.is_some(),
+            "current_address should be updated before send_batch call."
         );
-        Builder::new()
-            .name("solFwdStage".to_string())
-            .spawn(move || forwarding_stage.run())
-            .unwrap()
+        let conn = self
+            .connection_cache
+            .get_connection(&self.current_address.unwrap());
+        let _res = conn.send_data_batch_async(wire_transactions);
     }
+}
 
+pub(crate) fn spawn_forwarding_stage(
```

Comment:
Extracted `spawn_forwarding_stage` from `ForwardingStage::spawn` to reduce public interface of this module. Now `ForwardingStage` is private. Also this allows in the future to use this function to wrap creation of different non-vote clients.